### ISSUE TITLE
saved_object migrations: ensure progress can be made in presence of crashes

### ISSUE THREAD

**Issue by mhoffm-aiven** [opensearch-project/OpenSearch-Dashboards#3304]
**Is your feature request related to a problem? Please describe.**

Currently saved_object migration can get stalled if we crash and see a the index that we created before the crash after restart. The current protocol only checks if the new migrated index was already created; if we crash before the migration was completed any recovery will require manual intervention like stopping opensearch dashboards, deleting the index and retrying.

**Describe the solution you'd like**

I think we could improve the situation if we use a similar method like in this repository https://github.com/graup/es-distributed-lock to coordinate that only one opensearch dashboards instance performs the migrations. If it crashes, it will loose the lease and some other instance should be able to take over ( in case migrations are idempotent, i have not checked yet ). 

**Describe alternatives you've considered**

None so far.

**Comment by ananzh** [comment]
AI: Help team understand the issue. List action items if needed. If it needs more discussion, pls add it to HOT TOPIC in the triage note.

**Comment by mhoffm-aiven** [comment]
Hey! Thanks for your response, but sorry i didnt understand; can you please rephrase?

**Comment by abbyhu2000** [comment]
If application crashes before the saved_object migration is completed, the saved object migration will not be completed correctly. Users are still seeing the previous version saved object. 

Do we want to use only one opensearch dashboards instance to perform the migrations as the author suggested?

The work around solution is to just delete the incorrectly migrated saved object, and restart dashboard.

**Comment by mhoffm-aiven** [comment]
There is the manual work around, but it would be better if dashboards could handle that without operator intervention!

**Comment by joshuarrrr** [comment]
I agree that it would be better to have some automatic mitigation in these cases. @seraphjiang Do you have any perspective here on this proposed approach?

**Comment by AMoo-Miki** [comment]
@mhoffm-aiven I couldn't look at that repo because it didn't have any license on it.

Saved Object migration failures are a pain. We can certainly come up with a super lay way of making sure only one Dashboards instance is running the show using a simple lock + reasonable delay. However, more important is cleaning up the mess caused by a previous crash or failure that this instance or another instance caused.

In fact, I have never seen a crash in middle of migration but I have seen a ton of situations where migration failed and the mess was left behind which if were cleaned (like in a transaction) it could have been retried. It is possible that letting the mess remain was a technique to prevent continuous migrations from occurring - which is another thing we will need to handle.

Hopefully, someone will get a chance to think this through and come up with a solid plan.
